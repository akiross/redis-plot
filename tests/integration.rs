/// These are the integration tests for redis-plot. They will build the library and start
/// redis-server (which must be in PATH) loading that module.
/// Then, some tests are run: rendering outputs are tested using reference images stored
/// tests/snapshots. You can use the show_rle_snap python utility (in flake.nix) to display
/// the image for inspection.
use anyhow::Result;
use itertools::Itertools;
use redis::{Client, Commands};
use serde::Serialize;
use std::path::{Path, PathBuf};
use tempfile::tempdir;

/// Is a drop guard that will kill the redis-server when dropped.
struct ServerGuard(std::process::Child);

impl Drop for ServerGuard {
    fn drop(&mut self) {
        if let Err(_) = self.0.kill() {
            println!("Couldn't kill the server");
        }
        if let Err(_) = self.0.wait() {
            println!("Couldn't wait the server");
        }
    }
}

impl ServerGuard {
    /// Opens a new connection to redis-server, using localhost and default port.
    /// It will attempt connection up to 10 times in 10 seconds, then fail.
    fn get_connection() -> Result<redis::Connection> {
        let client = redis::Client::open("redis://127.0.0.1/").expect("Cannot connect to server");
        for i in 0..10 {
            match client.get_connection() {
                Ok(con) => {
                    // Connected, perform tests
                    return Ok(con);
                }
                Err(e) => {
                    if e.is_connection_refusal() {
                        // Wait for the server to be on line
                        std::thread::sleep_ms(1000);
                    } else {
                        // Cannot connect to server within a reasonable amount of time, fail.
                        return Err(e.into());
                    }
                }
            }
        }
        anyhow::bail!("Cannot connect to redis-server");
    }
}

/// This function builds the cdylib in a temporary directory, runs a callback
/// and returns its value removing the directory. If environment variable
/// REDIS_PLOT_TEST_TARGET_DIR is set, then that directory will be used as cargo target dir (which
/// might help for faster build times).
fn build_cdylib<F: FnOnce(&Path) -> Result<(), anyhow::Error>>(f: F) -> Result<(), anyhow::Error> {
    let temp_path = if let Some(p) = option_env!("REDIS_PLOT_TEST_TARGET_DIR") {
        PathBuf::from(p)
    } else {
        // Use a temporary directory for building the library, this is safer than using a fixed
        // directory, but will slow down build time during development.
        let temp_dir = tempdir().expect("Cannot create temporary directory");
        temp_dir.path().to_path_buf()
    };
    // TODO check building result: was cargo successful?
    std::process::Command::new("cargo")
        .args(["build", "--target-dir"])
        .arg(&temp_path)
        .output()
        .expect("Failed to build the library");
    let path = temp_path.join("debug").join("libredis_plot.so");
    // Ensure library was built and file is present
    assert!(path.exists());
    f(&path)
}

fn start_server(module_path: &Path) -> ServerGuard {
    // This is expected to be the path to the library
    // let module_path = env!("LIB_PATH");
    std::process::Command::new("redis-server")
        .arg("--loadmodule")
        .arg(module_path)
        .arg("--save") // Disable persistence, just noisy in tests.
        .arg("")
        .stdout(std::process::Stdio::null()) // TODO show output on failed tests
        .spawn()
        .map(ServerGuard)
        .expect("Cannot start redis-server")
}

/// RLE encoding of a slice.
fn rle(v: &[u8]) -> Vec<(usize, u8)> {
    // Apply RLE encoding to get something more manageable
    v.iter()
        .peekable()
        .batching(|it| {
            if let Some(first) = it.next() {
                // Count how many times we match the iterator
                let mut c: usize = 1;
                while let Some(n) = it.peek() {
                    if n == &first {
                        it.next();
                        c += 1;
                    } else {
                        // We are done finding this sequence
                        break;
                    }
                }
                Some((c, *first))
            } else {
                // No more elements in the iterator
                None
            }
        })
        .collect::<Vec<_>>()
}

#[test]
fn test_rle() {
    assert_eq!(rle(&vec![]), vec![]);
    assert_eq!(rle(&vec![0x01]), vec![(1, 0x01)]);
    assert_eq!(rle(&vec![0xff]), vec![(1, 0xff)]);
    assert_eq!(rle(&vec![0xff, 0xff, 0xff]), vec![(3, 0xff)]);
    assert_eq!(
        rle(&vec![0xff, 0xff, 0x00, 0x00, 0xff]),
        vec![(2, 0xff), (2, 0x00), (1, 0xff)]
    );
    assert_eq!(
        rle(&vec![0x11, 0x22, 0x33, 0x44, 0x55]),
        vec![(1, 0x11), (1, 0x22), (1, 0x33), (1, 0x44), (1, 0x55)]
    );
}

/// Represents a RLE-encoded RGB image and its shape, it's used by insta for
/// writing the snapshot file.
#[derive(Serialize)]
struct RleImage {
    width: usize,
    height: usize,
    rle_data: Vec<(usize, u8)>,
}

impl From<Vec<u8>> for RleImage {
    fn from(v: Vec<u8>) -> Self {
        Self {
            width: usize::from_be_bytes(v[..8].try_into().unwrap()),
            height: usize::from_be_bytes(v[8..16].try_into().unwrap()),
            rle_data: rle(&v[16..]),
        }
    }
}

/// Integration tests are placed here. There's currently a single test for
/// performance reasons,
#[test]
fn test_everything() -> Result<(), anyhow::Error> {
    use redis::Commands;
    build_cdylib(|lib_path| {
        let _server = start_server(lib_path);
        let mut con = ServerGuard::get_connection()?;

        // This is just a smoke test.
        assert_eq!(
            redis::cmd("rsp.echo").arg("foo").query(&mut con),
            Ok("rsp.echo, foo".to_owned())
        );

        // Build a few lists as plot targets
        let la: i32 = con.rpush("la", vec![1u32, 2, 4, 9, 15, 16, 42])?;
        assert_eq!(la, 7);
        let lb: i32 = con.rpush("lb", vec![-1, 1, -2, 2, -3, 3, -4, 4])?;
        assert_eq!(lb, 8);
        con.rpush("lc", vec![0.0, 0.5, 1.75, -2.125])?;
        con.rpush("ld", vec![0.125, 0, 0, 0, 125, 2.5, 3.0, 3.125, 2.95])?;

        // Test the rsp.draw command, which plots one list.
        {
            let res: Vec<u8> = redis::cmd("rsp.draw").arg("la").query(&mut con)?;
            insta::assert_json_snapshot!(RleImage::from(res));
        }
        // TODO plot two or more lists together
        // TODO plot one list as scatter, one as lines, one as histograms
        // TODO test async plotting: bind a list to a plot and try updating
        {
            // Bind a list to a plot
            redis::cmd("rsp.bind").arg("la").query(&mut con)?;
        }

        Ok(())
    })
}
