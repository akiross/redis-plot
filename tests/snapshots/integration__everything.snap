---
source: tests/integration.rs
expression: img
---
{
  "width": 160,
  "height": 90,
  "rle_data": [
    [
      14613,
      255
    ],
    [
      3,
      241
    ],
    [
      3,
      225
    ],
    [
      381,
      255
    ],
    [
      3,
      189
    ],
    [
      9,
      96
    ],
    [
      3,
      102
    ],
    [
      3,
      130
    ],
    [
      3,
      210
    ],
    [
      15,
      255
    ],
    [
      3,
      222
    ],
    [
      3,
      109
    ],
    [
      3,
      64
    ],
    [
      3,
      49
    ],
    [
      3,
      72
    ],
    [
      3,
      121
    ],
    [
      3,
      204
    ],
    [
      6,
      255
    ],
    [
      3,
      189
    ],
    [
      9,
      96
    ],
    [
      3,
      108
    ],
    [
      3,
      141
    ],
    [
      3,
      227
    ],
    [
      9,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      8
    ],
    [
      36,
      255
    ],
    [
      3,
      230
    ],
    [
      3,
      207
    ],
    [
      12,
      255
    ],
    [
      3,
      230
    ],
    [
      3,
      207
    ],
    [
      321,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      70
    ],
    [
      3,
      84
    ],
    [
      3,
      73
    ],
    [
      3,
      19
    ],
    [
      3,
      3
    ],
    [
      3,
      164
    ],
    [
      9,
      255
    ],
    [
      3,
      190
    ],
    [
      3,
      7
    ],
    [
      3,
      31
    ],
    [
      3,
      114
    ],
    [
      3,
      131
    ],
    [
      3,
      99
    ],
    [
      3,
      31
    ],
    [
      3,
      96
    ],
    [
      6,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      70
    ],
    [
      3,
      84
    ],
    [
      3,
      71
    ],
    [
      3,
      15
    ],
    [
      3,
      8
    ],
    [
      3,
      167
    ],
    [
      6,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      8
    ],
    [
      36,
      255
    ],
    [
      3,
      128
    ],
    [
      3,
      12
    ],
    [
      12,
      255
    ],
    [
      3,
      128
    ],
    [
      3,
      12
    ],
    [
      321,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      6,
      255
    ],
    [
      3,
      240
    ],
    [
      3,
      38
    ],
    [
      3,
      17
    ],
    [
      9,
      255
    ],
    [
      3,
      58
    ],
    [
      3,
      29
    ],
    [
      3,
      241
    ],
    [
      9,
      255
    ],
    [
      3,
      254
    ],
    [
      3,
      225
    ],
    [
      6,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      6,
      255
    ],
    [
      3,
      234
    ],
    [
      3,
      29
    ],
    [
      3,
      20
    ],
    [
      3,
      253
    ],
    [
      3,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      8
    ],
    [
      36,
      255
    ],
    [
      3,
      128
    ],
    [
      3,
      12
    ],
    [
      12,
      255
    ],
    [
      3,
      128
    ],
    [
      3,
      12
    ],
    [
      321,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      9,
      255
    ],
    [
      3,
      115
    ],
    [
      3,
      0
    ],
    [
      3,
      226
    ],
    [
      6,
      255
    ],
    [
      3,
      24
    ],
    [
      3,
      76
    ],
    [
      24,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      9,
      255
    ],
    [
      3,
      112
    ],
    [
      3,
      0
    ],
    [
      3,
      221
    ],
    [
      3,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      8
    ],
    [
      9,
      255
    ],
    [
      3,
      212
    ],
    [
      3,
      87
    ],
    [
      3,
      34
    ],
    [
      3,
      38
    ],
    [
      3,
      99
    ],
    [
      3,
      226
    ],
    [
      6,
      255
    ],
    [
      3,
      157
    ],
    [
      3,
      40
    ],
    [
      3,
      4
    ],
    [
      6,
      80
    ],
    [
      3,
      91
    ],
    [
      3,
      157
    ],
    [
      3,
      40
    ],
    [
      3,
      4
    ],
    [
      6,
      80
    ],
    [
      3,
      91
    ],
    [
      9,
      255
    ],
    [
      3,
      231
    ],
    [
      3,
      109
    ],
    [
      3,
      43
    ],
    [
      3,
      32
    ],
    [
      3,
      78
    ],
    [
      3,
      203
    ],
    [
      9,
      255
    ],
    [
      3,
      163
    ],
    [
      3,
      94
    ],
    [
      3,
      224
    ],
    [
      3,
      89
    ],
    [
      3,
      35
    ],
    [
      3,
      118
    ],
    [
      3,
      239
    ],
    [
      3,
      117
    ],
    [
      3,
      50
    ],
    [
      3,
      31
    ],
    [
      3,
      50
    ],
    [
      3,
      102
    ],
    [
      3,
      240
    ],
    [
      237,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      9,
      255
    ],
    [
      3,
      89
    ],
    [
      3,
      2
    ],
    [
      3,
      238
    ],
    [
      6,
      255
    ],
    [
      3,
      63
    ],
    [
      3,
      14
    ],
    [
      3,
      192
    ],
    [
      21,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      9,
      255
    ],
    [
      3,
      100
    ],
    [
      3,
      0
    ],
    [
      3,
      228
    ],
    [
      3,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      8
    ],
    [
      6,
      255
    ],
    [
      3,
      210
    ],
    [
      3,
      13
    ],
    [
      3,
      43
    ],
    [
      3,
      148
    ],
    [
      3,
      140
    ],
    [
      3,
      26
    ],
    [
      3,
      25
    ],
    [
      3,
      231
    ],
    [
      3,
      255
    ],
    [
      3,
      193
    ],
    [
      3,
      72
    ],
    [
      3,
      7
    ],
    [
      6,
      144
    ],
    [
      3,
      151
    ],
    [
      3,
      193
    ],
    [
      3,
      72
    ],
    [
      3,
      7
    ],
    [
      6,
      144
    ],
    [
      3,
      151
    ],
    [
      6,
      255
    ],
    [
      3,
      228
    ],
    [
      3,
      28
    ],
    [
      3,
      37
    ],
    [
      3,
      142
    ],
    [
      3,
      157
    ],
    [
      3,
      70
    ],
    [
      3,
      9
    ],
    [
      3,
      201
    ],
    [
      6,
      255
    ],
    [
      3,
      120
    ],
    [
      3,
      13
    ],
    [
      3,
      34
    ],
    [
      3,
      101
    ],
    [
      3,
      148
    ],
    [
      3,
      170
    ],
    [
      3,
      72
    ],
    [
      3,
      29
    ],
    [
      3,
      143
    ],
    [
      3,
      170
    ],
    [
      3,
      146
    ],
    [
      3,
      76
    ],
    [
      3,
      224
    ],
    [
      237,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      179
    ],
    [
      3,
      216
    ],
    [
      3,
      205
    ],
    [
      3,
      135
    ],
    [
      3,
      3
    ],
    [
      3,
      89
    ],
    [
      9,
      255
    ],
    [
      3,
      199
    ],
    [
      3,
      17
    ],
    [
      3,
      0
    ],
    [
      3,
      26
    ],
    [
      3,
      86
    ],
    [
      3,
      154
    ],
    [
      3,
      233
    ],
    [
      9,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      6,
      255
    ],
    [
      3,
      189
    ],
    [
      3,
      11
    ],
    [
      3,
      40
    ],
    [
      6,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      8
    ],
    [
      6,
      255
    ],
    [
      3,
      73
    ],
    [
      3,
      20
    ],
    [
      3,
      240
    ],
    [
      6,
      255
    ],
    [
      3,
      218
    ],
    [
      3,
      5
    ],
    [
      3,
      107
    ],
    [
      6,
      255
    ],
    [
      3,
      128
    ],
    [
      3,
      12
    ],
    [
      12,
      255
    ],
    [
      3,
      128
    ],
    [
      3,
      12
    ],
    [
      15,
      255
    ],
    [
      3,
      91
    ],
    [
      3,
      29
    ],
    [
      3,
      241
    ],
    [
      9,
      255
    ],
    [
      3,
      57
    ],
    [
      3,
      68
    ],
    [
      6,
      255
    ],
    [
      3,
      120
    ],
    [
      3,
      0
    ],
    [
      3,
      146
    ],
    [
      6,
      255
    ],
    [
      3,
      251
    ],
    [
      3,
      4
    ],
    [
      3,
      153
    ],
    [
      252,
      255
    ],
    [
      3,
      148
    ],
    [
      15,
      0
    ],
    [
      3,
      91
    ],
    [
      3,
      249
    ],
    [
      12,
      255
    ],
    [
      3,
      234
    ],
    [
      3,
      141
    ],
    [
      3,
      72
    ],
    [
      3,
      18
    ],
    [
      3,
      0
    ],
    [
      3,
      14
    ],
    [
      3,
      172
    ],
    [
      6,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      20
    ],
    [
      3,
      24
    ],
    [
      3,
      12
    ],
    [
      3,
      0
    ],
    [
      3,
      39
    ],
    [
      3,
      204
    ],
    [
      6,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      8
    ],
    [
      3,
      255
    ],
    [
      3,
      251
    ],
    [
      3,
      8
    ],
    [
      3,
      107
    ],
    [
      12,
      255
    ],
    [
      3,
      73
    ],
    [
      3,
      35
    ],
    [
      6,
      255
    ],
    [
      3,
      128
    ],
    [
      3,
      12
    ],
    [
      12,
      255
    ],
    [
      3,
      128
    ],
    [
      3,
      12
    ],
    [
      12,
      255
    ],
    [
      3,
      253
    ],
    [
      3,
      12
    ],
    [
      3,
      105
    ],
    [
      3,
      240
    ],
    [
      6,
      239
    ],
    [
      3,
      238
    ],
    [
      3,
      127
    ],
    [
      3,
      7
    ],
    [
      3,
      254
    ],
    [
      3,
      255
    ],
    [
      3,
      120
    ],
    [
      3,
      0
    ],
    [
      3,
      244
    ],
    [
      9,
      255
    ],
    [
      3,
      27
    ],
    [
      3,
      54
    ],
    [
      3,
      191
    ],
    [
      3,
      251
    ],
    [
      246,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      179
    ],
    [
      3,
      216
    ],
    [
      3,
      196
    ],
    [
      3,
      80
    ],
    [
      3,
      6
    ],
    [
      3,
      190
    ],
    [
      21,
      255
    ],
    [
      3,
      253
    ],
    [
      3,
      183
    ],
    [
      3,
      25
    ],
    [
      3,
      13
    ],
    [
      3,
      236
    ],
    [
      3,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      126
    ],
    [
      3,
      152
    ],
    [
      3,
      164
    ],
    [
      3,
      196
    ],
    [
      3,
      252
    ],
    [
      9,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      8
    ],
    [
      3,
      255
    ],
    [
      3,
      233
    ],
    [
      3,
      0
    ],
    [
      3,
      140
    ],
    [
      12,
      255
    ],
    [
      3,
      107
    ],
    [
      3,
      10
    ],
    [
      6,
      255
    ],
    [
      3,
      128
    ],
    [
      3,
      12
    ],
    [
      12,
      255
    ],
    [
      3,
      128
    ],
    [
      3,
      12
    ],
    [
      12,
      255
    ],
    [
      3,
      234
    ],
    [
      24,
      0
    ],
    [
      3,
      242
    ],
    [
      3,
      255
    ],
    [
      3,
      120
    ],
    [
      3,
      17
    ],
    [
      12,
      255
    ],
    [
      3,
      193
    ],
    [
      3,
      38
    ],
    [
      3,
      0
    ],
    [
      3,
      11
    ],
    [
      3,
      78
    ],
    [
      3,
      188
    ],
    [
      240,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      6,
      255
    ],
    [
      3,
      252
    ],
    [
      3,
      47
    ],
    [
      3,
      33
    ],
    [
      3,
      249
    ],
    [
      24,
      255
    ],
    [
      3,
      147
    ],
    [
      3,
      0
    ],
    [
      3,
      185
    ],
    [
      3,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      21,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      8
    ],
    [
      3,
      255
    ],
    [
      3,
      242
    ],
    [
      3,
      0
    ],
    [
      3,
      125
    ],
    [
      12,
      255
    ],
    [
      3,
      92
    ],
    [
      3,
      20
    ],
    [
      6,
      255
    ],
    [
      3,
      128
    ],
    [
      3,
      12
    ],
    [
      12,
      255
    ],
    [
      3,
      128
    ],
    [
      3,
      12
    ],
    [
      12,
      255
    ],
    [
      3,
      242
    ],
    [
      3,
      0
    ],
    [
      3,
      120
    ],
    [
      18,
      240
    ],
    [
      6,
      255
    ],
    [
      3,
      120
    ],
    [
      3,
      20
    ],
    [
      18,
      255
    ],
    [
      3,
      213
    ],
    [
      3,
      145
    ],
    [
      3,
      52
    ],
    [
      3,
      0
    ],
    [
      3,
      188
    ],
    [
      237,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      9,
      255
    ],
    [
      3,
      186
    ],
    [
      3,
      0
    ],
    [
      3,
      149
    ],
    [
      24,
      255
    ],
    [
      3,
      144
    ],
    [
      3,
      0
    ],
    [
      3,
      186
    ],
    [
      3,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      21,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      8
    ],
    [
      6,
      255
    ],
    [
      3,
      35
    ],
    [
      3,
      62
    ],
    [
      9,
      255
    ],
    [
      3,
      252
    ],
    [
      3,
      32
    ],
    [
      3,
      68
    ],
    [
      6,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      6
    ],
    [
      12,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      6
    ],
    [
      15,
      255
    ],
    [
      3,
      42
    ],
    [
      3,
      57
    ],
    [
      24,
      255
    ],
    [
      3,
      120
    ],
    [
      3,
      20
    ],
    [
      24,
      255
    ],
    [
      3,
      250
    ],
    [
      3,
      15
    ],
    [
      3,
      114
    ],
    [
      237,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      12,
      255
    ],
    [
      3,
      57
    ],
    [
      3,
      29
    ],
    [
      3,
      249
    ],
    [
      3,
      255
    ],
    [
      3,
      73
    ],
    [
      3,
      157
    ],
    [
      3,
      232
    ],
    [
      3,
      255
    ],
    [
      3,
      252
    ],
    [
      3,
      186
    ],
    [
      3,
      23
    ],
    [
      3,
      22
    ],
    [
      3,
      245
    ],
    [
      3,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      21,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      8
    ],
    [
      6,
      255
    ],
    [
      3,
      143
    ],
    [
      3,
      0
    ],
    [
      3,
      143
    ],
    [
      3,
      253
    ],
    [
      3,
      250
    ],
    [
      3,
      111
    ],
    [
      3,
      0
    ],
    [
      3,
      176
    ],
    [
      6,
      255
    ],
    [
      3,
      166
    ],
    [
      3,
      0
    ],
    [
      3,
      169
    ],
    [
      3,
      230
    ],
    [
      3,
      234
    ],
    [
      3,
      255
    ],
    [
      3,
      166
    ],
    [
      3,
      0
    ],
    [
      3,
      169
    ],
    [
      3,
      230
    ],
    [
      3,
      234
    ],
    [
      6,
      255
    ],
    [
      3,
      167
    ],
    [
      3,
      0
    ],
    [
      3,
      110
    ],
    [
      3,
      242
    ],
    [
      3,
      255
    ],
    [
      3,
      253
    ],
    [
      3,
      203
    ],
    [
      3,
      156
    ],
    [
      6,
      255
    ],
    [
      3,
      120
    ],
    [
      3,
      20
    ],
    [
      9,
      255
    ],
    [
      3,
      235
    ],
    [
      3,
      144
    ],
    [
      3,
      235
    ],
    [
      6,
      255
    ],
    [
      3,
      195
    ],
    [
      3,
      3
    ],
    [
      3,
      150
    ],
    [
      237,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      12,
      255
    ],
    [
      3,
      184
    ],
    [
      3,
      0
    ],
    [
      3,
      149
    ],
    [
      3,
      255
    ],
    [
      3,
      88
    ],
    [
      3,
      8
    ],
    [
      3,
      0
    ],
    [
      3,
      10
    ],
    [
      3,
      3
    ],
    [
      3,
      0
    ],
    [
      3,
      38
    ],
    [
      3,
      200
    ],
    [
      6,
      255
    ],
    [
      3,
      148
    ],
    [
      3,
      0
    ],
    [
      3,
      212
    ],
    [
      21,
      255
    ],
    [
      3,
      132
    ],
    [
      3,
      8
    ],
    [
      6,
      255
    ],
    [
      3,
      254
    ],
    [
      3,
      108
    ],
    [
      3,
      2
    ],
    [
      3,
      12
    ],
    [
      3,
      8
    ],
    [
      3,
      6
    ],
    [
      3,
      134
    ],
    [
      9,
      255
    ],
    [
      3,
      247
    ],
    [
      3,
      72
    ],
    [
      6,
      0
    ],
    [
      3,
      16
    ],
    [
      3,
      255
    ],
    [
      3,
      247
    ],
    [
      3,
      72
    ],
    [
      6,
      0
    ],
    [
      3,
      16
    ],
    [
      9,
      255
    ],
    [
      3,
      149
    ],
    [
      3,
      14
    ],
    [
      3,
      2
    ],
    [
      3,
      24
    ],
    [
      3,
      5
    ],
    [
      3,
      4
    ],
    [
      3,
      111
    ],
    [
      6,
      255
    ],
    [
      3,
      120
    ],
    [
      3,
      20
    ],
    [
      9,
      255
    ],
    [
      3,
      230
    ],
    [
      3,
      14
    ],
    [
      3,
      0
    ],
    [
      3,
      20
    ],
    [
      3,
      24
    ],
    [
      3,
      2
    ],
    [
      3,
      87
    ],
    [
      3,
      251
    ],
    [
      237,
      255
    ],
    [
      3,
      243
    ],
    [
      3,
      224
    ],
    [
      3,
      251
    ],
    [
      15,
      255
    ],
    [
      3,
      225
    ],
    [
      3,
      234
    ],
    [
      6,
      255
    ],
    [
      3,
      246
    ],
    [
      3,
      200
    ],
    [
      6,
      174
    ],
    [
      3,
      209
    ],
    [
      12,
      255
    ],
    [
      3,
      243
    ],
    [
      3,
      224
    ],
    [
      3,
      251
    ],
    [
      21,
      255
    ],
    [
      3,
      241
    ],
    [
      3,
      225
    ],
    [
      12,
      255
    ],
    [
      3,
      225
    ],
    [
      3,
      174
    ],
    [
      3,
      177
    ],
    [
      3,
      233
    ],
    [
      18,
      255
    ],
    [
      3,
      243
    ],
    [
      6,
      226
    ],
    [
      9,
      255
    ],
    [
      3,
      243
    ],
    [
      6,
      226
    ],
    [
      12,
      255
    ],
    [
      3,
      244
    ],
    [
      3,
      188
    ],
    [
      3,
      170
    ],
    [
      3,
      191
    ],
    [
      3,
      236
    ],
    [
      9,
      255
    ],
    [
      3,
      239
    ],
    [
      3,
      227
    ],
    [
      12,
      255
    ],
    [
      3,
      252
    ],
    [
      3,
      205
    ],
    [
      3,
      176
    ],
    [
      3,
      180
    ],
    [
      3,
      222
    ],
    [
      3168,
      255
    ],
    [
      243,
      0
    ],
    [
      238,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      1,
      255
    ],
    [
      2,
      0
    ],
    [
      18312,
      255
    ]
  ]
}
